runtime/platform/platform.cpp

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include <executorch/runtime/platform/log.h>
+#include <executorch/runtime/platform/platform.h>
+#include <cstdlib>
+
+namespace executorch {
+namespace runtime {
+
+namespace {
+/**
+ * The singleton instance of the PAL function table.
+ */
+pal_table pal_impl = pal_table(
+    et_pal_init,
+    et_pal_abort,
+    et_pal_current_ticks,
+    et_pal_ticks_to_ns_multiplier,
+    et_pal_emit_log_message,
+    et_pal_allocate,
+    et_pal_free,
+    __FILE__);
+
+/**
+ * Tracks whether the PAL has been overridden. This is used to warn when
+ * multiple callers override the PAL.
+ */
+bool is_pal_overridden = false;
+} // namespace
+
+pal_table::pal_table(
+    et_pal_emit_log_message_t emit_log_message,
+    const char* source_filename)
+    : emit_log_message(emit_log_message), source_filename(source_filename) {}
+
+pal_table::pal_table(
+    et_pal_init_t init,
+    et_pal_abort_t abort,
+    et_pal_current_ticks_t current_ticks,
+    et_pal_ticks_to_ns_multiplier_t ticks_to_ns_multiplier,
+    et_pal_emit_log_message_t emit_log_message,
+    et_pal_allocate_t allocate,
+    et_pal_free_t free,
+    const char* source_filename)
+    : init(init),
+      abort(abort),
+      current_ticks(current_ticks),
+      ticks_to_ns_multiplier(ticks_to_ns_multiplier),
+      emit_log_message(emit_log_message),
+      allocate(allocate),
+      free(free),
+      source_filename(source_filename) {}
+
+/**
+ * Override the PAL functions with user implementations. Any null entries in the
+ * table are unchanged and will keep the default implementation.
+ */
+bool register_pal(pal_table impl) {
+  if (is_pal_overridden) {
+    ET_LOG(
+        Error,
+        "register_pal() called multiple times. Subsequent calls will override the previous implementation. Previous implementation was registered from %s.",
+        impl.source_filename != nullptr ? impl.source_filename : "unknown");
+  }
+  is_pal_overridden = true;
+
+  if (impl.abort != nullptr) {
+    pal_impl.abort = impl.abort;
+  }
+
+  if (impl.current_ticks != nullptr) {
+    pal_impl.current_ticks = impl.current_ticks;
+  }
+
+  if (impl.ticks_to_ns_multiplier != nullptr) {
+    pal_impl.ticks_to_ns_multiplier = impl.ticks_to_ns_multiplier;
+  }
+
+  if (impl.emit_log_message != nullptr) {
+    pal_impl.emit_log_message = impl.emit_log_message;
+  }
+
+  if (impl.allocate != nullptr) {
+    pal_impl.allocate = impl.allocate;
+  }
+
+  if (impl.free != nullptr) {
+    pal_impl.free = impl.free;
+  }
+
+  if (impl.init != nullptr) {
+    pal_impl.init = impl.init;
+    if (pal_impl.init != nullptr) {
+      pal_impl.init();
+    }
+  }
+
+  return true;
```

Comment:
assert the whole table is non-nullptr?

Reply:
This should be statically impossible, since we initialize to the default implementations and only override with non-null entries, but I'm not opposed to adding the assert if desired.